### ISSUE TITLE
Enable easy and customizable inlining of html / javascript / css to initial page 

### ISSUE THREAD

**Issue by pleku** [vaadin/flow#3010]

**Comment by Legioth** [comment]
This should already work automatically for any dependency annotation with `LoadMode.INLINE` on any component that is present during the initial rendering. If needed, the same functionality can also be used together with whatever mechanism is used for #3011.

**Comment by pleku** [comment]
The problem with `LoadMode.INLINE` is that it is not currently usable for all the cases and even the most common cases it seems.

For `@JavaScript`, it currently inlines everything **by appending** to `<head>`, but there is a use case for prepending them instead in Bakery.

For `@HtmlImport`, it currently inlines everything to ´<body>`, which is not desired if you want to add stuff to `<head>` instead, e.g. inline all the favicon definitions or a `<custom-style>`, like the bakery does.

I would force it to an explicit decision to the user to either define `LoadMode` as 
- `INLINE_HEAD_PREPEND` - added before flow stuff
- `INLINE_HEAD_APPEND`- added after flow stuff
- `INLINE_BODY` - this would always prepend since appending is the same as just using `LAZY` (?)

Another option would be to use a separate annotation for these, but I'm a bit against that since I don't see why (oh why).

**Comment by Legioth** [comment]
I'm starting to consider the opposite alternative of using a different mechanism for inlining contents. The reason for this is that the `LoadMode.INLINE` is confusing as it's only fully appropriate for loading the initial page (although we currently try to emulate the same semantics for subsequently discovered annotations), but it's applied in a context that isn't tied to the initial page load.

I'd instead propose removing `LoadMode.INLINE` and instead introduce a separate `@Inline` annotation that is only usable in the same context as e.g. the proposed `@Viewport` annotation.

`@Inline` would have these parameters:
* `String value` is the classpath location (i.e. not messing with any `frontend://` protocols or such) of the file to inline
* `WrappingMode wrappingMode` defines whether the file should be inlined as-is (i.e. as HTML), or wrapped in e.g. a `<style>` or `<script>` tag. The default setting selects the mode to use based on the filename extension (and throws at deployment time for unsupported extensions).
* `InlinePosition inlinePosition` defines where to inline the content, e.g. the beginning or end of either `<head>` or `<body>`. By default, content is inlined into the end of `<head>`.

**Comment by pleku** [comment]
What @Legioth [described](https://github.com/vaadin/flow/issues/3010#issuecomment-346548455), except we'll split the `InlinePosition` to `Target` (head/body) and `Position`(beginning/end or prepend/append ?) 

### Acceptance Criteria
- There documentation with examples on how to inline content for different parts of the initial page

**Comment by Legioth** [comment]
I guess there should also be a programmatic API for doing the same things in a `PageConfigurator`?

**Comment by pleku** [comment]
> I guess there should also be a programmatic API for doing the same things in a PageConfigurator?

Yes, at least it could be, but does it have to if this is enough? 
Anyways not part of this ticket, there will be moar tickets to the Epic #3011 coming

**Comment by SomeoneToIgnore** [comment]
I would like to notice that `flow-maven-plugin` needs to be adjusted to look for new annotations also and propose either to create a new ticket for this or do not close this one, until it's done.

**Comment by caalador** [comment]
@SomeoneToIgnore The annotations are read on runtime so no specific lookup is needed at any point as the accepted files are only read from the classpath.

**Comment by SomeoneToIgnore** [comment]
Maven plugin is not able to get anything from Flow in runtime, it can only statically analyze the sources.
It needs to get all the html , css, js files to process them later: minify, transpile, bundle.

**Comment by caalador** [comment]
I understand, but we won't be using anything bundled for the initial page setting. It's always inllined so we need to actually read the file and it's then added into an element.

**Comment by pleku** [comment]
Yeah this has nothing to do with maven plugin and bundling.

Transpilation is another thing, the user might want to inline different resources pending on the loaded browser. For this we need the dynamic API, as Leif already asked after
>I guess there should also be a programmatic API for doing the same things in a PageConfigurator?

Created https://github.com/vaadin/flow/issues/3139 for this
